boofuzz/sessions.py

```diff
@@ -363,7 +371,9 @@ def __init__(self, session_filename=None, index_start=1, index_end=None, sleep_t
         self._receive_data_after_fuzz = receive_data_after_fuzz
         self._skip_current_node_after_current_test_case = False
         self._skip_current_element_after_current_test_case = False
-        self._post_test_case_methods = []
+        self._pre_send_methods = pre_send_callback
+        self._post_test_case_methods = post_test_case_callback
+        self._restart_methods = restart_callback
```

Comment:
Should these be lists? They seem to be treated as iterable later on.

Reply:
They should be yes, but I thought checking for None is a little more performant than calling len().
The main reason I went for this approach is, that setting a default value to [] triggers a warning in PyCharm. I can't remember which one exactly, will look it up.

Reply:
So this is the soft warning you get in PyCharm when setting default values to list type.
`Inspection info: This inspection detects when a mutable value as list or dictionary is detected in a default value for an argument. Default argument values are evaluated only once at function definition time, which means that modifying the default value of the argument will affect all subsequent calls of the function.`
The argument stated there isn't really valid for an empty list as that value won't be modified.
So using [] should be possible in this case but currently I don't get why setting them to None would be an issue or let's say not best practice.

Reply:
> I thought checking for None is a little more performant than calling len().
The performance difference should be fairly negligible compared to what else happens in a fuzz test. See https://softwareengineering.stackexchange.com/questions/80084/is-premature-optimization-really-the-root-of-all-evil

> setting a default value to [] triggers a warning in PyCharm
This is a good reason, and the warning is valid. It's definitely a language corner case. For an extended discussion, see: https://stackoverflow.com/questions/9158294/good-uses-for-mutable-function-argument-default-values

Now that I look at it again, I see you documented that the arguments to Session should be lists themselves:
```
        pre_send_callback (list of method): The registered method will be called prior to each fuzz request.
                                            Default None.
```

This would be clearer if the argument names were plural. `pre_send_callbacks` instead of `pre_send_callback`, etc.

As for the default value situation, the way one can resolve this is to keep the default value at None, and add some code inside `__init__`:
```
        if pre_send_callbacks is None:
            pre_send_callbacks = []
        self._pre_send_methods = pre_send_callbacks
```

This makes the constructor more bulky, but has the significant advantage of giving maintainers one less thing to worry about -- `self._pres_send_methods` is always a list. No checks for None.

Requested changes:
1. Pluralize Session argument names.
2. (If you have time to change and test) Initialize callback lists to empty lists by default, and remove the None checks later in the code.